File: neetcode all and misc questions/Dynamic Programming and Recursion/2D Dynamic Programming/edit_distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        """
        https://www.youtube.com/watch?v=fJaKO8FbDdo&ab_channel=takeUforward
        Approach:
        Starting from the back and working towards the beginning of both strings,
        defining i as length of word1 and j as the length of word2
        
        At each point we have three different operations
        insert
        replace
        delete
        
        we want to run recursion and find the minimum
        between the three operations
        
        When defining the recurrence relation, we think about comparing the characters
        of both strings at points i, j and how the operations affect this
        
        insert:
        assuming i = 4
        j = 2
        horse
        ros
        
        say if we insert s 
        horses
        ros
        
        i remains at 4, because we've now matched the hypothetical character "s"
        that we just inserted with the "s" at j = 2, we can now shift by one to the left
        
        therefore, insert = f(i, j-1)
        
        delete:
        assuming i = 4
        j = 2
        horse
        ros
        
        if we delete e, we still need to match "s", so i shifts by one and j stays the same
        
        in this case,
        i is now 3, and s now matches
        therefore, delete = f(i-1, j)
        
        replace:
        assume i = 4
        j = 2
        horse
        ros
        
        if we replace e with s,
        we need to shift both indices i and j, since i and j both match, so we can move onto the remainder of the string
        
        Base Cases:
        
        If we've decreased the index to the point where word1 is now an empty string, and there are still characters for word2,
        it takes len(word2) insert operations to take an empty string to word2, therefore:
        
        return j + 1
        
        since j is length of word2 - 1, so we add back the 1
        
        if we've decreased the index to the point where word2 is now an empty string, and there are still characters for word1,
        it takes len(word1 insert operations to take an empty string to word1, therefore
        
        return i + 1
        
        since j is length of word1 - 1, so we add back the 1
        
        If both the characters at word1 and word2 are the same, then no operations are needed, but we shift both i and j by one
        
        return f(i-1, j-1)
       
       	-------------------------------

       	We can then use memoization to store our results as (i, j) to cut down
       	on repeated subproblems
	
		Without memoization, the time complexity would be exponential (3^N)
       	Final Time Complexity: O(N*M)
       	Space Complexity: O(N*M)
        """
        self.memo = dict()
        def f(i, j, word1, word2):
            if i < 0:
                return j + 1
            if j < 0:
                return i + 1
            if (i,j) in self.memo:
                return self.memo[(i,j)]
            if word1[i] == word2[j]:
                self.memo[(i,j)] = f(i-1, j-1, word1, word2)
                return self.memo[(i,j)]
            insert = f(i, j-1, word1, word2)
            replace = f(i-1,j-1, word1, word2)
            delete = f(i-1,j, word1, word2)
            # in our recurrence relation, we want to find the operation that results in the minimal
            # amount of total operations needed, so we take the min between these three
            # the addition of 1 represents that at this particular step, we need to do at least one operation.
            self.memo[(i,j)] = 1 + min(insert, replace, delete)
            return self.memo[(i,j)]
        
        return f(len(word1)-1,len(word2)-1, word1, word2)
        

File: node_js_ds_and_algos_implementations/2D dynamic programming/edit_distance.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
/*
https://leetcode.com/problems/edit-distance/discuss/5241237/java-or-dp-or-top-down
recurrence relation:

when comparing word1 and word2, we think about
the relation between the indices i, j, and what the operation
allows us to increment in order to go further into the recursion

insert:
"horse"
"ros"

if we were to insert "r" to get "rhorse" and "ros", this would
mean that we've successfully matched "r" with this hypothetical character that we just added.
However, because we just added a char, we have to keep i the same, since this "increased" the 
length of the string, so we still need to evaluate the char at word1[i] (in this case "h")
f(i, j+1)

replace:
"horse"
"ros"

if we were to replace "h" with "r", we'd increment both i and j,
since we've successfully matched both characters
f(i+1, j+1)

delete:
"horse"
"ros"

if we were to delete "h", we'd get "orse" and "ros"
in this example, we didn't match any character, but because we haven't matched
a letter in word2 yet, we increment i but keep j the same

f(i+1, j)

Since the question asks for minimum number of operations, we want to take the min of all three operations
and return this. We'd also add one to this result to show that an operation was made at this recursive call
to add onto the min number of operations that we'll be returning in each call.

Edge Case:
if at any point, word1[i] === word2[j],
this means that we can also increment both i and j

Base Case:
if we reach the end of a word, we need to figure out how many remaining operations would be needed to 
get word1 to word2

if end of word1 is reached,
    the number of necessary operations would be the difference in length between word2 and j 
if end of word2 is reached,
    the number of necessary operations would be the difference in length between word1 and i 

Memoization:
at each (i, j), we store the min amount of operations that it'd take to reach word2, given the remaining characters
starting from i, j

Time Complexity (with memoization):
O(M*N), where M is the length of word1 and N is the length of word2
The reasoning is because even with memoization, for each letter of word1, the recursion still takes us through 
through each letter of word2 (Although some of these calls will be O(1) due to retrieving a previously calculated result
from the memoization dict)
Space:
O(M*N)
*/
var minDistance = function(word1, word2) {
    let M = word1.length
    let N = word2.length
    let memo = {}
    var search = function(i, j){
        if (j === N){
            return M-i
        }
        if (i === M){
            return N-j
        }
        if (`${i},${j}` in memo){
            return memo[`${i},${j}`]
        }
        if (word1[i] === word2[j]){
            memo[`${i},${j}`] = search(i+1,j+1)
            return memo[`${i},${j}`]
        }
        let insert = 1 + search(i, j+1)
        let replace = 1 + search(i+1, j+1)
        let del = 1 + search(i+1, j)
        memo[`${i},${j}`] = Math.min(insert, replace, del)
        return memo[`${i},${j}`]
    }
    return search(0, 0)
}; 